httpx/concurrency/trio.py

```diff
@@ -0,0 +1,241 @@
+import functools
+import math
+import ssl
+import typing
+from types import TracebackType
+
+import trio
+
+from ..config import PoolLimits, TimeoutConfig
+from ..exceptions import ConnectTimeout, PoolTimeout, ReadTimeout, WriteTimeout
+from .base import (
+    BaseBackgroundManager,
+    BaseEvent,
+    BasePoolSemaphore,
+    BaseQueue,
+    BaseStream,
+    ConcurrencyBackend,
+    TimeoutFlag,
+)
+
+
+def _or_inf(value: typing.Optional[float]) -> float:
+    return value if value is not None else float("inf")
+
+
+class Stream(BaseStream):
+    def __init__(self, stream: trio.abc.Stream, timeout: TimeoutConfig) -> None:
+        self.stream = stream
+        self.timeout = timeout
+        self.is_eof = False
+        self.write_buffer = b""
+        self.write_lock = trio.Lock()
+
+    def get_http_version(self) -> str:
+        if not isinstance(self.stream, trio.SSLStream):
+            return "HTTP/1.1"
+
+        ident = self.stream.selected_alpn_protocol()
+        if ident is None:
+            ident = self.stream.selected_npn_protocol()
+
+        return "HTTP/2" if ident == "h2" else "HTTP/1.1"
+
+    async def read(
+        self, n: int, timeout: TimeoutConfig = None, flag: TimeoutFlag = None
+    ) -> bytes:
+        if timeout is None:
+            timeout = self.timeout
+
+        while True:
+            # Check our flag at the first possible moment, and use a fine
+            # grained retry loop if we're not yet in read-timeout mode.
+            should_raise = flag is None or flag.raise_on_read_timeout
+            read_timeout = _or_inf(timeout.read_timeout if should_raise else 0.01)
+
+            with trio.move_on_after(read_timeout):
+                data = await self.stream.receive_some(max_bytes=n)
+                # b"" is the expected EOF message for Trio.
+                # The other case is an edge case that occurs with uvicorn+httptools.
+                if data == b"" or data.endswith(b"0\r\n\r\n"):
```

Comment:
The `b"0\r\n\r\n"` is the last chunk of a `Transfer-Encoding: chunked` response. I don't think we should be handling this at this level, this is an HTTP construct.

Reply:
Yes, that’s what I eventually figured that. The difficulty here is that Trio doesn’t expose an equivalent for asyncio’s stream_reader.at_eof(). And since we only call read() when we receive a NEED_DATA event we are never given the chance to receive an empty bytes message... So the question is: how can we otherwise detect that we received the whole response?

Reply:
(Also the comment here is plain wrong. 😅)

Reply:
cc @njsmith :)

Reply:
It sounds like there's something fundamentally confused here... whenever you get `NEED_DATA` you call `receive_some` and feed what you get to h11. You know you've received the whole response when h11 reports an `EndOfMessage` event.

The one small wrinkle is that when `receive_some` reports EOF, you need to tell h11 that you've gotten an EOF, you need to tell h11 that you've gotten an EOF. Conveniently, the way you do this is by calling `h11_conn.receive_data(b"")`.

Reply:
Thanks for chiming in @njsmith.

> when receive_some reports EOF

The thing is — `.receive_some()` never actually reports an EOF, neither on `TrioBackend` nor `AsyncioBackend`. My proof is a modest print statement on the return value of `backend.read()` while running:

```bash
HTTPX_DEBUG=1 pytest -k test_connection_closed_free_semaphore_on_acquire -s -m asyncio
HTTPX_DEBUG=1 pytest -k test_connection_closed_free_semaphore_on_acquire -s -m trio
```

As far as I understand the current `HTTP11Connection` implementation, when we receive an `EndOfMessage` in `._receive_response_data()` we simply `break` out of the response body iterator. As a result, `response.close()` is called, and the next h11 cycle is prepared. Then the connection is put in a keep-alive state, and we're mostly done.

So apparently, never in this teardown process do we call an equivalent of `h11_state.receive_data(b"")`, and I think that makes sense because that would result in a `ConnectionClosed` event, but the connection has *not* been closed.

With this in mind, I truly wonder how come the asyncio `stream_reader` detects the EOF at all. The following raw setup is that it does detect an EOF when receiving empty bytes as well, just like trio:

- Uvicorn server:

```python
async def app(scope, receive, send):
    await send(
        {
            "type": "http.response.start",
            "status": 200,
            "headers": [[b"content-type", b"text/plain"]],
        }
    )
    await send({"type": "http.response.body", "body": b"Hello, world!"})
```

- Client script:

```python
import asyncio
import trio

from httpx.concurrency.asyncio import AsyncioBackend
from httpx.concurrency.trio import TrioBackend

from httpx.config import TimeoutConfig, DEFAULT_TIMEOUT_CONFIG


async def main(backend):
    print(f"Going to connect using {backend!r}")
    input("Please start the server, then continue.")
    stream = await backend.connect(
        "localhost",
        8000,
        ssl_context=None,
        timeout=TimeoutConfig(DEFAULT_TIMEOUT_CONFIG),
    )

    assert not stream.is_connection_dropped()
    input("Please close the server, then continue.")

    # This is the interesting part.
    assert not stream.is_connection_dropped()
    assert await stream.read(1) == b""
    assert stream.is_connection_dropped()

    await stream.close()
    print("Done.")


asyncio.run(main(AsyncioBackend()))
trio.run(main, TrioBackend())
```

But like I said, we never seem to `.read()` empty bytes at all in HTTPX, but for some reason `stream_reader.at_eof()` returns `True` when trying to acquire a new connection after the server restarted. Maybe `stream_reader.at_eof()` does extra processing that I don't know about?

Anyway, I'm really confused, indeed. @tomchristie Any clues on a missing part of the read/write flow I'm missing here?

Reply:
asyncio always keeps reading from the socket in the background, whether you call `read` or not, as long as the loop is running. Trio and sync code don't. Unfortunately, httpx currently relies on the asyncio behavior to detect when an idle connection has gone stale. IMO this is a bug in httpx's idle connection handling, but that's the current state – there's some more discussion in this thread, in this comment and the next few below it: https://github.com/encode/httpx/pull/143#issuecomment-515152590

You'll have to ask Tom how he wants to deal with it here...

Reply:
Thanks a lot @njsmith. 🥇 From reading the thread you linked to, I learnt about `.is_readable()`. It turns out to exactly what I need for detecting whether the connection has dropped at the time of re-acquiring it.

I opened a separate issue (#327) for correctly handling connection resets in the asyncio case. So, I guess this is fixed?

Reply:
Hmm, sorry about that, I was returning `.is_readable()` instead of `not .is_readable()` in `is_connection_dropped()`. 🤦‍♂ 

@njsmith The commit I pushed is here: https://github.com/encode/httpx/pull/276/commits/bea9592217ec98c88769ef715414a94ff56bf55e

So now `is_readable()` returns `True` even after having restarted the server. I'm not sure why that is, but there's probably a missing step because HTTPX does seem to be relying on asyncio's background socket management — at least that's probably the reason why we only check for dropped connections once, i.e. before acquiring a new one.

Reply:
Oh right, I actually misremembered the issue we were discussing in #143 – that was about the distinction between "closed" and "readable", which is much more subtle. So arguably it's a bug that when we want to detect if a connection is reusable, then the asyncio backend checks for EOF, when ideally it should be checking for readability. But the way your patch handles it looks right to me.